connections/stream-migration.md

```diff
@@ -0,0 +1,238 @@
+already closed, it should also close the new stream for writing. Specifically
+imagine the following case.
+
+
+![stream-migration-half-closed](./stream-migration/stream-migration-half-closed.svg)
+
+<details>
+  <summary>Instructions to reproduce diagram</summary>
+``` plantuml
+@startuml stream-migration-half-closed
+skinparam sequenceMessageAlign center
+entity Initiator
+entity Responder
+
+Initiator <- Responder: <b>Stream A:</b> ""EOF""
+note over Responder: <b>Stream A</b> is closed for writing
+
+== Stream Migration ==
+
+note over Initiator, Responder: Migrate <b>Stream A</b> to <b>Stream B</b>
+
+Initiator -> Responder: Open new stream on <b>Connection 2</b>. Call this <b>Stream B</b>
+
+Initiator -> Responder: <b>Stream B:</b> Negotiate stream-migration protocol with ""<stream-migration protocol id>""
+Initiator -> Responder: <b>Stream B:</b> Send ""StreamMigration(type=Migrate, id=B, from=A)"" message
+
+Initiator <- Responder: <b>Stream B:</b> Ack Migrate
+
+note over Initiator
+    Close <b>stream A</b> for writing.
+    Will only write to <b>stream B</b> from now on.
+end note
+
+note over Initiator
+    We have already seen the ""EOF"" on
+    <b>Stream A</b> from <i>Responder</i>
+    So we continue reading on <b>stream B</b>
+end note
+
+Initiator -> Responder: <b>Stream A:</b> ""EOF""
+
+note over Responder
+    Treat ""EOF"" on <b>stream A</b> as a signal to close <b>stream A</b> for
+    writing and continue writing on <b>stream B</b>. However stream A was
+    already closed (before migration), so we close <b>stream B</b> as well here.
+end note
+Initiator <- Responder: <b>Stream B:</b> ""EOF""
+
+note over Initiator, Responder: Stream A is now migrated to Stream B
+
+@enduml
+```
+To generate:
+```bash
+plantuml stream-migration.md -o stream-migration -tsvg
+```
+</details>
+
+The reverse case where the Initiator's stream is closed for writing is the same
+as above, but mirrored.
+
+## Picking the best connection
+
+Moving streams from one connection to another involves picking which connection
+we should move the streams to. Here are some recommended heuristics the
+initiator may use in determining which connection is best.
+
+1. If we have both relayed and direct connections, keep the direct connections
+   and drop the relay connections.
+2. Check for simultaneous connect: If we have both inbound and outbound
+   connections, keep the ones initiated by the peer with the lowest peer ID. Open
+   Question: Some protocols behave differently depending on whether they are the
+   dialer or listener. Can we really consolidate these?
+3. Prefer the connection with the most streams.
+4. Break ties in the remaining connections by selecting the newest conn, to
+   match the swarm's behavior in best connection selection.
+
+Note that it's not required that all implementations (and all versions) follow
+the same heuristics since the initiator is driving the migration and specifies
```

Comment:
Below I am assuming that *Initiator* refers to the *connection* initiator, not the stream initiator. Please correct me in case I am misunderstanding this @MarcoPolo.

Say there are two nodes `A` and `B`. Connection `AB` is initiated by `A` to `B`. Conneciton `BA` is initiated by `B` to `A`.

Say that `A` and `B` follow different heuristics to pick the best connection. `A` chooses `AB` as the best connection, `B` chooses `BA` as the best connection.

If I understand the above correctly, this would result in `A` moving all the streams it created to `AB` and `B` moving all the streams it created to `BA`. Both connections would thus stay alive.

Potential solution: Instead of allowing both `A` and `B` to migrate streams, how about delegating the decision making to the peer with the lower peer ID, e.g. in this case `A`?

Reply:
Yes!

I was thinking about this this weekend and came up with a similar solution. Glad to see you also came to the same conclusion. I'll update this spec to make this explicit.

Reply:
Few thoughts on this:
1. Currently each peer chooses its own IDs for streams, i.e. there are two distinct spaces of stream IDs. If we want to allow the receiver of a stream to migrate that stream, we need a single stream ID space. One way to realize this would be to mandate the client (roles as seen by the stream muxer) to use odd and the server to use even stream IDs.
2. I don't think this document should describe how peers would choose `AB` over `BA`. This document should only describe how to migrate one libp2p stream from one muxer stream to another. For all that this spec cares about, those streams might (or might not) live on the same underlying connection. We can _then_ use the stream migration protocol as a building block to converge onto one connection (and the peer ID comparison is quite a neat idea, I like it!), but that should probably be described in a different document.

Reply:
1. Yes. I was considering adding a boolean flag that would indicate which peer initially identified the stream it's referencing when migrating (was I the initiator of the from stream?). This is the same as using even and odd numbers, since that scheme effectively encodes this boolean in the least significant bit. I'm fine with either way. Maybe it's a little easier to think about even and odds, so I'll do that.

2. Agreed that describing how to sort connections is out of the scope of this document (I imagine that spec to iterate more and and possibly have more subtle details). But I do think this spec should define who is responsible for doing the stream migration. If we end up in the situation where we have two identical connections (`A` dialed `B` and `B` dialed `A` at roughly the same time) we should describe who is in charge of doing the stream migration. By defining which node starts the stream migration we simplify this protocol and also avoid having to handle cases where both sides start stream migration at the same time.

Reply:
"Potential solution: Instead of allowing both A and B to migrate streams, how about delegating the decision making to the peer with the lower peer ID, e.g. in this case A?"

Wouldn't this create a biais toward lower peerIDs? Maybe we can hash the concatenation of the two peerIDs (the lower first). If the hash is even, use the lowest. Else use the highest. That way it is deterministic but no peerID is systematically favored over another.

Reply:
Seems reasonable enough, I added that. Thanks!

Reply:
> Wouldn't this create a biais toward lower peerIDs? Maybe we can hash the concatenation of the two peerIDs (the lower first). If the hash is even, use the lowest. Else use the highest. That way it is deterministic but no peerID is systematically favored over another.

What would be the upside/downside for a node to be "systematically favored"? Also note that an attacker could still pre-generate its peer ID in order to have the lower or higher one.

An attacker could try to force streams of a remote onto a slower connection. Though I doubt this would (a) have a big impact and (b) be the only way to do so.

Reply:
I don't think there's any security protection that this gives you. The one thing it does give you is a avoiding the case where a peer happens to be in the lower end of the key space and thus always has to handle this responsibility. With the hash approach (and cooperative nodes) peers have a 50/50 odds of handling this responsibility, so it should be more uniform.

Reply:
Yeah on second thought I agree with Max. I actually don't see the benefits since it's still 50% odds either way.

Reply:
I'm still not convinced that we should specify anything here at all. Stream migration is a general feature, a building block.

The use case we have in mind _now_ is migrating all streams from one connection to another, but we might come up with other use cases in the future. I'd prefer to have stream migration just be a thing that any node, regardless of its peer ID, can use _in principle_.
For the _particular_ use case of converging on a single connection, comparing peer IDs seems reasonable.

Reply:
Let me try to rephrase to see if I understand:

1. The protocol should only specify how a node would perform a stream migration.
2. It doesn't define who starts the migration or why.
3. Consolidating connections would be a layer on top of this that defines which nodes is in charge of migrating streams to empty and close connections.

We just have to make sure that what we design here doesn't block point 3.

If that seems accurate, then I agree we don't need this in here.

Reply:
also @marten-seemann to highlight a recent change re:

> Currently each peer chooses its own IDs for streams, i.e. there are two distinct spaces of stream IDs. If we want to allow the receiver of a stream to migrate that stream, we need a single stream ID space. One way to realize this would be to mandate the client (roles as seen by the stream muxer) to use odd and the server to use even stream IDs.

I've specced something similar, except the lower peer id node uses even and the higher peer id node uses odds. This let's us avoid having to rely on the stream muxer to give us this role. And it also works across connections (it gets confusing if the stream muxer says we are the client on connection and the server on the other).